src/Http/Middleware/DoublePassMiddleware.php

```diff
@@ -0,0 +1,58 @@
+<?php
+declare(strict_types=1);
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         4.0.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Http\Middleware;
+
+use Cake\Core\Exception\Exception;
+use Cake\Http\Response;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+
+class DoublePassMiddleware implements MiddlewareInterface
+{
+    protected $callable;
+
+    public function __construct(callable $callable)
+    {
+        $this->callable = $callable;
+    }
+
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler) : ResponseInterface
+    {
+        return $response = ($this->callable)(
+            $request,
+            new Response(),
+            $this->decorateHandler($handler)
+        );
+    }
+
+    public function getCallable(): callable
+    {
+        return $this->callable;
+    }
+
+    protected function decorateHandler(RequestHandlerInterface $handler): callable
+    {
+        return function ($request, $response) use ($handler) {
+            try {
+                return $handler->handle($request);
+            } catch (Exception $e) {
+                return $response;
```

Comment:
I think `Runner` could require `$fallbackHandler` in constructor as in [Queue-based request handler](https://www.php-fig.org/psr/psr-15/meta/#queue-based-request-handler) from PSR-15 Meta Document.

Currently in both usages, CakePHP adds last item to queue:
* `Server::run()` adds `Application` to queue as last item.
* `RoutingMiddleware` also adds `$next` to queue as last item.
* Even tests uses same technique.

This would allow to avoid exhausted queue and runner will always be able handle request.

Reply:
Nice, my bad for not reading the meta document.

In practice we are using the fallback approach already as you pointed out, just need to formalize the API accordingly.

The queue exhaustion issue was primary an issue in tests only.

Reply:
I chose a simple approach and now the `Runner` returns a prototype response if the queue is exhausted.